echopype/calibrate/calibrate_ek.py

```diff
@@ -309,18 +309,18 @@ def _get_chan_dict(beam: xr.Dataset) -> Dict:
         """
         # Use center frequency for each ping to select BB or CW channels
         # when all samples are encoded as complex samples
-        if "frequency_start" in beam and "frequency_end" in beam:
-            # At least some channels are BB
-            # frequency_start and frequency_end are NaN for CW channels
-            freq_center = (beam["frequency_start"] + beam["frequency_end"]) / 2
+        if not np.all(beam["transmit_type"] == "CW"):
+            # At least 1 BB ping exists -- this is analogous to what we had from before
+            # Before: when at least 1 BB ping exists, frequency_start and frequency_end will exist
+            # transmit_frequency_start and transmit_frequency_stop are NaN for CW channels
+            freq_center = (beam["transmit_frequency_start"] + beam["transmit_frequency_stop"]) / 2
 
             return {
-                # For BB: drop channels containing CW samples (nan in freq start/end)
-                "BB": freq_center.dropna(dim="channel").channel,
-                # For CW: drop channels containing BB samples (not nan in freq start/end)
-                "CW": freq_center.where(np.isnan(freq_center), drop=True).channel,
+                # For BB: Keep only non-CW channels (FM or FMD) based on transmit_type
+                "BB": freq_center.where(beam["transmit_type"] != "CW", drop=True).channel,
+                # For CW: Keep only CW channels based on transmit_type
+                "CW": freq_center.where(beam["transmit_type"] == "CW", drop=True).channel,
             }
```

Comment:
We do not need `freq_center` here anymore now that `transmit_type` is available.

The `BB: ...` and `CW: ...` selections do not work if there are within-channel changes of CW and FM. In those cases, either only the single elements fits the criteria are kept, or only the single elements fits the criteria are set to nan. For example, if you have such within-channel changes:
<img width="517" alt="Screenshot 2023-08-02 at 5 36 33 PM" src="https://github.com/OSOceanAcoustics/echopype/assets/15334215/53ef1d23-1dfa-428d-b5ca-c2f9bd543758">
Then the following happens:
first selection:
<img width="523" alt="Screenshot 2023-08-02 at 5 37 54 PM" src="https://github.com/OSOceanAcoustics/echopype/assets/15334215/6dd95555-2ea6-4be4-84c6-6b4dd408fe57">
second selection:
<img width="456" alt="Screenshot 2023-08-02 at 5 36 48 PM" src="https://github.com/OSOceanAcoustics/echopype/assets/15334215/3b00fe17-7809-48aa-9c1a-9e16ed95d284">

I think we can do the following
```python
first_ping_transmit_type = beam["transmit_type"].isel(ping_time=0)  # assume transmit_type identical for all pings
return {
    # For BB: Keep only non-CW channels (FM or FMD) based on transmit_type
    "BB": first_ping_transmit_type.where(first_ping_transmit_type != "CW", drop=True).channel,
    # For CW: Keep only CW channels based on transmit_type
    "CW": first_ping_transmit_type.where(first_ping_transmit_type == "CW", drop=True).channel,
}
```

Reply:
I follow the logic and it seems like it'd work. But I ran into multiple errors when testing it. I'll have to go into debug mode on it tomorrow morning, to look at it more carefully.

Reply:
Alright, I found and fixed the problem. The first statement in your suggestion retained the `ping_time` variable, which needs to be removed. So, the fixed statement is:
```python
first_ping_transmit_type = beam["transmit_type"].isel(ping_time=0).drop_vars("ping_time")
```